jsonrpc-specs/src/methods/discv5.json

```diff
@@ -0,0 +1,567 @@
+					},
+					"recipientPort": {
+						"description": "recipient UDP port",
+						"type": "number"
+					}
+				}
+			}
+		}
+	},
+	{
+		"name": "discv5_findNodes",
+		"summary": "Send a FINDNODES request for nodes that fall within the given set of distances, to the designated peer and wait for a response.",
+		"params": [
+			{
+				"name": "node_id",
+				"required": true,
+				"schema": {
+					"title": "Ethereum node record",
+					"$ref": "#/components/schemas/Enr"
+				}
+			},
+			{
+				"name": "distances",
+				"required": true,
+				"schema": {
+					"title": "distance",
+					"type": "array",
+					"items": {
+						"type": "number"
+					}
+				}
+			}
+		],
+		"result": {
+			"name": "findNodesResult",
+			"description": "Returns nodes in a given distance",
+			"schema": {
+				"title": "FOUNDNODES message",
+				"type": "array",
+				"items": { "$ref": "#/components/schemas/Enr" }
+			}
+		}
+	},
+	{
+		"name": "discv5_talk",
+		"summary": "Send a TALKREQUEST request with a payload to a given peer and wait for response.",
+		"params": [
+			{
+				"name": "node_id",
+				"required": true,
+				"schema": {
+					"title": "Ethereum node record",
+					"$ref": "#/components/schemas/Enr"
+				}
+			},
+			{
+				"name": "protocol",
+				"required": true,
+				"description": "protocol element of message-data",
+				"schema": {
+					"title": "protocol id",
+					"$ref": "#/components/schemas/hexString"
+				}
+			},
+			{
+				"name": "payload",
+				"required": true,
+				"description": "request element of message-data",
+				"schema": {
+					"$ref": "#/components/schemas/hexString"
+				}
+			}
+		],
+		"result": {
+			"name": "talkResult",
+			"description": "Returns TALKRESPONSE message as hex string",
+			"schema": {
+				"$ref": "#/components/schemas/hexString"
+			}
+		}
+	},
+	{
+		"name": "discv5_recursiveFindNodes",
+		"summary": "Lookup a target node within in the network",
+		"params": [
+			{
+				"name": "node_id",
+				"required": true,
+				"schema": {
+					"title": "Ethereum node record",
+					"$ref": "#/components/schemas/Enr"
+				}
+			}
+		],
+		"result": {
+			"name": "recursiveFindNodesResult",
+			"schema": {
+				"type": "array",
+				"items": { "$ref": "#/components/schemas/Enr" }
+			}
+		}
+	},
+	{
+		"name": "discv5_bond",
```

Comment:
I understand this comes from https://ddht.readthedocs.io/en/latest/jsonrpc.html, but it is not fully clear to me what this is supposed to do, as it is also not mentioned in other specifications.

Does it just send a ping (then why not use ping?), or does it send some repetitive ping?

Reply:
@pipermerriam could you elaborate please on the intention of this API call?

Reply:
Yes, it was from this:

https://github.com/ethereum/ddht/blob/142911d134ff839f3f79ff8fe9e45d3fe5a58cd0/ddht/v5_1/network.py#L392-L434

It encapsulated the following logic:

- liveliness check via PING/PONG
- ensure latest ENR record for remote node
- add to routing table if there is room

The reason for this abstraction was:

- I didn't implement automatic adding to routing table for outbound round trip PING/PONG messages because... I'm not sure I had a good reason but it didn't feel like it should be bundled at the time I wrote it.
- I needed a liveliness check that also took care of the additional book-keeping of making sure ENR was updated so that I after "bonding" a section of code could be confident that they could communicate with a node.

I'm not confident that this needs to be in the JSON-RPC spec.  I will encourage you guys to consider driving development of the specification by using it to find out what is needed and what is superfluous.  Writing down a set of use cases for externally driving the client via JSON-RPC and then evaluating what methods will be needed to implement that functionality will probably go a long way towards understanding what methods are useful and which ones are not.

For example:

* Use Case:
    * Check that a remote node adds us to their routing table when they receive an inbound ping
* Functionality we need:
    * Spin up Alice, Bob, and Carol such that:
        * Carol is considered a "bootnode" and has no configured bootnodes.
        * Alice is configured with Carol as a bootnode so that Alice will initially connect to Carol which satisfies the requirement that Alice's routing table is populated.  Otherwise, eclipse attack protection mechanisms would prevent her from adding Bob when Bob sends a ping.
        * Bob is configured with no bootnodes
    * Check that Alice's routing table contains Carol and **does not** contain Bob
    * Initiate a PING from Bob to Alice
        * Ideally we can also confirm that Bob receives the PONG response from Alice
    * Check that Alice is now in Bob's routing table
    * Check that Bob is now in Alice's routing table
* Specific JSON-RPC functionality we need
    * Ability to start a node with control over the list of bootnodes
    * Ability to view full routing table of a node
    * Ability to perform a PING/PONG round trip between two nodes.

Reply:
Yes, for this reason we should probably add JSON-RPC functionality as they become needed by our (testing) use cases.